tree/20_cfe_basics/time_lib.cf

```diff
@@ -0,0 +1,216 @@
+    delay_too_small.!delay_null::
+      "Warning, max_execution_delay for ${job_id} is less than your agent run period but not null, it will be ignored";
+}
+
+
+# Simple Periodic Scheduler, it raises a class on a regular perioridicity, based on a user friendly definition.
+#
+# job_id: a string to identify this job
+# agent_periodicity: how often you run the agent in minutes
+# max_execution_delay_minutes, max_execution_delay_hours: on how much time you want to spread the job
+# start_on_minutes, start_on_hours, start_on_day_of_week: when should be the first run
+# periodicity_minutes, periodicity_hours, periodicity_days: how often should the job run
+# mode: "nodups": avoid duplicate runs in the same period
+#       "catchup": avoid duplicates and one or more run have been missed, run once before next period
+#       "stateless": no check is done on past runs
+bundle agent schedule_simple_generic(job_id, agent_periodicity,
+                                     max_execution_delay_minutes, max_execution_delay_hours,
+                                     start_on_minutes, start_on_hours, start_on_day_of_week,
+                                     periodicity_minutes, periodicity_hours, periodicity_days, mode)
+{
+  vars:
+    any::
+      # max_execution_delay in minutes
+      "max_execution_delay_real" string => eval("${max_execution_delay_minutes} + 60 * ${max_execution_delay_hours}", "math", "infix");
+      "max_execution_delay" int => "${max_execution_delay_real}";
+
+      # periodicity in minutes
+      "job_periodicity_real" string => eval("${periodicity_minutes} + 60 * ${periodicity_hours} + 60 * 24 * ${periodicity_days}", "math", "infix");
+      "job_periodicity" int => "${job_periodicity_real}";
+
+      # compute time between start_* and now
+      # (min_of_day - start_minutes) + 60 * (hour of day - start hours) + 60 * 24 * (day of week - start day)%7 
+      "minutes_from_start_real" string => eval(strftime("localtime", "%M - ${start_on_minutes} + 60 * (%H - ${start_on_hours}) + 60 * 24 * ((%u - ${start_on_day_of_week} + 7)%%7)", "${ncf_date.timestamp}"), "math", "infix");
+      "minutes_from_start" int => "${minutes_from_start_real}";
+
+  classes:
+    any::
+      "nodups" expression => strcmp("${mode}", "nodups");
+      "catchup" expression => strcmp("${mode}", "catchup");
+
+    pass2.!catchup::
+      "should_run" expression => strcmp("${schedule_generic.current_run_id}", "${schedule_generic.job_run_id}");
+    pass2.catchup::
+      "should_run" not => islessthan("${schedule_generic.current_run_id}", "${schedule_generic.job_run_id}");
+    pass2::
+      "job_${job_id}_run" expression => "should_run.!job_${job_id}_persist.!job_${job_id}_error",
+                          scope => "namespace";
+    pass2.(nodups|catchup)::
+      "job_${job_id}_persist" expression => "job_${job_id}_run",
+                              persistence => "${job_periodicity}"; # shouldn't there be a -1 here ?
```

Comment:
It was a question for you too: if we have an interval of 5mn and we persist 5 minutes, will the next run be canceled because it is withing those 5mn (with a few second error margin) ?